### PR TITLE
[TOOLS-1988] Linking Touch Time by Issues with Timeline for filtering.

### PR DESCRIPTION
**Pull request by marcokuchla**
Adds start and end progressing timestamps to send to FE to be able to filter by timeline;

### CODE REVIEW

File: dashboard/src/main/resources/static/scripts/dc-utils.js [hunk 1]
```diff
@@ -215,90 +215,90 @@ function DcUtils() {
...
+
+    // this.insertPlotDataIntoChart = function (chart, plotData, layerNamesKey) {
+    //     const xScale = d3.time.scale().domain([plotData.startDate, plotData.endDate]);
+    //     chart.x(xScale);
+    //     chart.dimension(plotData.dimension);
         
-        function stackAccessor (layerName) {
-            return (item) => item.value[layerName];
-        }
+    //     function stackAccessor (layerName) {
```
Comment (objective-solutions-user) [inline dashboard/src/main/resources/static/scripts/dc-utils.js]: MAJOR Remove this commented out code. [rule](http://10.44.1.250:9000/coding_rules#rule_key=javascript%3ACommentedCode)

File: dashboard/src/main/resources/static/scripts/dc-utils.js [hunk 2]
```diff
@@ -215,90 +215,90 @@ function DcUtils() {
...
+    //         .legend(this.getDefaultLegend())
+    //         .colors(this.getDefaultColors())
+    //         .transitionDuration(0)
+    //         // mouse interaction
+    //         .brushOn(false) // don't select with drag-n-drop
+    //         .renderHorizontalGridLines(true)
+    //         .elasticY(true);
+    // };
+
+    // this.reverseLegendOrder = function(chart) {
```
Comment (objective-solutions-user) [inline dashboard/src/main/resources/static/scripts/dc-utils.js]: MAJOR Remove this commented out code. [rule](http://10.44.1.250:9000/coding_rules#rule_key=javascript%3ACommentedCode)

File: dashboard/src/main/resources/static/scripts/dc-utils.js [hunk 3]
```diff
@@ -215,90 +215,90 @@ function DcUtils() {
...
-            }
-            p[groupingKey] -= v[valuePropertyNameKey];
-            return p;
-        };
-        const reduceInit = () => {
-            return {};
-        };
-        crossfilterGroup.reduce(reduceAdd, reduceSub, reduceInit);
-    }
+    //     const tooltipAccessor = function (item) {
```
Comment (objective-solutions-user) [inline dashboard/src/main/resources/static/scripts/dc-utils.js]: MAJOR Remove this commented out code. [rule](http://10.44.1.250:9000/coding_rules#rule_key=javascript%3ACommentedCode)

File: dashboard/src/main/resources/static/scripts/dc-utils.js [hunk 4]
```diff
@@ -215,90 +215,90 @@ function DcUtils() {
...
-            const items = chart.__legendables();
-            return items.reverse();
-        };
-    };
-
-    this.insertPlotDataIntoChart = function (chart, plotData, layerNamesKey) {
-        const xScale = d3.time.scale().domain([plotData.startDate, plotData.endDate]);
-        chart.x(xScale);
-        chart.dimension(plotData.dimension);
+    // this.configureDefaultStackedChart = function(chart) {
```
Comment (objective-solutions-user) [inline dashboard/src/main/resources/static/scripts/dc-utils.js]: MAJOR Remove this commented out code. [rule](http://10.44.1.250:9000/coding_rules#rule_key=javascript%3ACommentedCode)

File: dashboard/src/main/resources/static/scripts/dc-utils.js [hunk 5]
```diff
@@ -215,90 +215,90 @@ function DcUtils() {
...
+    //             // first stack group must be added as group
+    //             chart.group(plotData.dateGroup, layerName, stackAccessor(layerName));
+    //         } else {
+    //             chart.stack(plotData.dateGroup, layerName, stackAccessor(layerName));
+    //         }
+    //         chart.title(layerName, tooltipAccessor);
+    //     }
+    // };
+
+    // this.createPaperListFilter = function (div, dimension, promptValue, promptText) {
```
Comment (objective-solutions-user) [inline dashboard/src/main/resources/static/scripts/dc-utils.js]: MAJOR Remove this commented out code. [rule](http://10.44.1.250:9000/coding_rules#rule_key=javascript%3ACommentedCode)

**Comment by objective-solutions-user** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 5 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: core/src/main/resources/static/scripts/chart-utils.js [hunk 6]
```diff
@@ -372,4 +394,33 @@ class TouchTimeChartBuilder extends ChartBuilderBase {
...
 }
+
+class Chart {
+    constructor (chart, filterCallback) {
+        this.chart = chart;
+        this.chartKey = chart.options.chart.renderTo;
+        this._filterCallback = filterCallback;
+    }
+
+    notify () {
```
Comment (herbert-venancio) [inline core/src/main/resources/static/scripts/chart-utils.js]: `notify` what? It's too generic. Maybe `notifyValueChanged`, `selectionChanged`, `dateRangeChanged`, `timelineChanged` would be more interesting.

File: kpis/src/main/resources/static/elements/taskboard/kpis/widget-touch-time.html [hunk 7]
```diff
@@ -192,51 +192,60 @@ <h3 class="tb-label">Level</h3>
...
-                    const series = this._transformToHighchartsSeries(effortByStatusByIssueKey, this._plotData.issueKeys);
+                    const series = this._transformToHighchartsSeries(effortByStatusByIssueKey, selectedIssueKeys);
+
                     this.chart = new TouchTimeChartBuilder('touchTime')
                         .withChartType('bar')
                         .withTooltipNumberOfDecimals(2)
-                        .withCategories(this._plotData.issueKeys)
+                        .withCategories(selectedIssueKeys)
                         .withSeriesData(series)
+                        .withTimelineFilterCallback(this._createChart.bind(this))
```
Comment (herbert-venancio) [inline kpis/src/main/resources/static/elements/taskboard/kpis/widget-touch-time.html]: Does it mean that a new chart is created every time? Don't you have to destroy the old one when it happens?

File: kpis/src/test/java/objective/taskboard/followup/kpi/enviroment/KPIEnvironmentBuilder.java [hunk 8]
```diff
@@ -67,6 +71,11 @@ public KPIEnvironmentBuilder withKpiProperties(KPIProperties kpiProperties) {
         this.kpiProperties = kpiProperties;
         return this;
     }
+    
+    public KPIEnvironmentBuilder setNow(Instant now) {
```
Comment (herbert-venancio) [inline kpis/src/test/java/objective/taskboard/followup/kpi/enviroment/KPIEnvironmentBuilder.java]: If you accept a string, you can do the repetitive `parseDateTime(today).toInstant()` calls here

File: kpis/src/test/java/objective/taskboard/followup/kpi/touchTime/TouchTimeKPIDataProviderTest.java [hunk 9]
```diff
@@ -367,6 +373,16 @@ public TouchTimePointsAsserter and() {
                 }
                 return this;
             }
+
+            public TouchTimePointsAsserter hasStartProgressingDate(String date) {
+                assertThat(currentPoint.startProgressingDate, is(Date.from(parseDateTime(date).toInstant())));
```
Comment (herbert-venancio) [inline kpis/src/test/java/objective/taskboard/followup/kpi/touchTime/TouchTimeKPIDataProviderTest.java]: no need to `Date.from` if comparing `Instant`

File: basic-utilities/pom.xml [hunk 10]
```diff
@@ -19,5 +19,15 @@
...
+		<dependency>
+			<groupId>junit</groupId>
+			<artifactId>junit</artifactId>
+			<scope>test</scope>
+		</dependency>
+		<dependency>
+			<groupId>org.assertj</groupId>
+			<artifactId>assertj-core</artifactId>
+			<scope>test</scope>
+		</dependency>
```
Comment (herbert-venancio) [inline basic-utilities/pom.xml]: Add this too:
```
		<dependency>
			<groupId>org.codehaus.sonar-plugins.java</groupId>
			<artifactId>sonar-jacoco-listeners</artifactId>
			<scope>test</scope>
		</dependency>
```

Comment (marcokuchla) [reply basic-utilities/pom.xml]: Added.

File: kpis/src/main/java/objective/taskboard/followup/kpi/touchTime/TouchTimeDataPoint.java [hunk 11]
```diff
@@ -1,15 +1,22 @@
...
 class TouchTimeDataPoint {
     public final String issueKey;
     public final String issueType;
     public final String issueStatus;
     public final double effortInHours;
+    public final Instant startProgressingDate;
+    public final Instant endProgressingDate;
     
-    public TouchTimeDataPoint(String issueKey, String issueType, String issueStatus, double effortInHours) {
+    public TouchTimeDataPoint(String issueKey, String issueType, String issueStatus, double effortInHours, ZonedDateTime startProgressingDate, ZonedDateTime endProgressingDate) {
```
[the remark this answers is not in the document]
Comment (marcokuchla) [reply kpis/src/main/java/objective/taskboard/followup/kpi/touchTime/TouchTimeDataPoint.java]: done.

File: kpis/src/main/java/objective/taskboard/followup/kpi/touchTime/TouchTimeKPIDataProvider.java [hunk 12]
```diff
@@ -35,18 +38,26 @@ public TouchTimeChartDataSet getDataSet(String projectKey, KpiLevel kpiLevel, Zo
...
         final List<String> statuses = kpiLevel.filterProgressingStatuses(allProgressingStatuses, jiraProperties);
-        final List<TouchTimeDataPoint> points = transformToDataPoints(issues, statuses);
+        final List<TouchTimeDataPoint> points = transformToDataPoints(issues, statuses, timezone);
         return new TouchTimeChartDataSet(points);
     }
 
-    private List<TouchTimeDataPoint> transformToDataPoints(List<IssueKpi> issues, List<String> statuses) {
+    private List<TouchTimeDataPoint> transformToDataPoints(List<IssueKpi> issues, List<String> statuses, ZoneId timezone) {
         final List<TouchTimeDataPoint> points = new LinkedList<>();
         issues.stream().forEach(issue -> {
```
[the remark this answers is not in the document]
Comment (marcokuchla) [reply kpis/src/main/java/objective/taskboard/followup/kpi/touchTime/TouchTimeKPIDataProvider.java]: Done.

File: basic-utilities/src/test/java/objective/taskboard/config/converter/TaskboardJacksonModuleTest.java [hunk 13]
```diff
@@ -0,0 +1,44 @@
+package objective.taskboard.config.converter;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.util.Date;
```
[the remark this answers is not in the document]
Comment (marcokuchla) [reply basic-utilities/src/test/java/objective/taskboard/config/converter/TaskboardJacksonModuleTest.java]: Fixed.

File: core/src/main/resources/static/scripts/chart-utils.js [hunk 14]
```diff
@@ -372,4 +394,33 @@ class TouchTimeChartBuilder extends ChartBuilderBase {
...
+
+    withTimelineFilterCallback (callback) {
+        this._filterCallback = callback;
+        return this;
+    }
+
+    build () {
+        const highchartChart = super.build();
+        if (this._filterCallback === undefined) {
+            throw 'FilterCallback must be defined';
```
[the remark this answers is not in the document]
Comment (marcokuchla) [reply core/src/main/resources/static/scripts/chart-utils.js]: Fixed.

File: core/src/main/resources/static/scripts/chart-utils.js (hunk 6 above)
Comment (marcokuchla) [reply core/src/main/resources/static/scripts/chart-utils.js]: Changed to notifyTimelineChanged().

File: kpis/src/main/resources/static/elements/taskboard/kpis/widget-touch-time.html (hunk 7 above)
Comment (marcokuchla) [reply kpis/src/main/resources/static/elements/taskboard/kpis/widget-touch-time.html]: Yes, but the charts library has some internal mechanisms to keep control and it destroys the old reference. Even though the library does that magic, I've created a new function to explicity call destroy before creating a new chart.

File: kpis/src/main/resources/static/elements/taskboard/kpis/widget-touch-time.html [hunk 15]
```diff
@@ -192,51 +192,60 @@ <h3 class="tb-label">Level</h3>
...
+                _transformToHighchartsSeries: function (effortByStatusByIssueKey, issueKeysOrderLUT) {
+                    const series = effortByStatusByIssueKey.map(
                         ({key: status, value: effortByIssueKey}) => {
                             return {
                                 name: status, 
                                 data: issueKeysOrderLUT.map((issueKey) => effortByIssueKey[issueKey])
                             };
                         }
                     );
+                    return series;
```
[the remark this answers is not in the document]
Comment (marcokuchla) [reply kpis/src/main/resources/static/elements/taskboard/kpis/widget-touch-time.html]: Fixed.

File: kpis/src/test/java/objective/taskboard/followup/kpi/enviroment/KPIEnvironmentBuilder.java (hunk 8 above)
Comment (marcokuchla) [reply kpis/src/test/java/objective/taskboard/followup/kpi/enviroment/KPIEnvironmentBuilder.java]: Overloaded the method to receive a String and updated where it was used to pass only the String;

File: kpis/src/test/java/objective/taskboard/followup/kpi/touchTime/TouchTimeKPIDataProviderTest.java (hunk 9 above)
Comment (marcokuchla) [reply kpis/src/test/java/objective/taskboard/followup/kpi/touchTime/TouchTimeKPIDataProviderTest.java]: Fixed.

File: kpis/src/main/java/objective/taskboard/followup/kpi/IssueKpi.java [hunk 16]
```diff
@@ -24,12 +24,14 @@
     private Optional<IssueTypeKpi> issueType;
     private List<IssueKpi> children = new LinkedList<>();
     private KpiLevel level;
+    private Clock clock;
 
-    public IssueKpi(String pKey, Optional<IssueTypeKpi> issueType, KpiLevel level, Optional<StatusTransition> firstStatus) {
+    public IssueKpi(String pKey, Optional<IssueTypeKpi> issueType, KpiLevel level, Optional<StatusTransition> firstStatus, Clock clock) {
```
[the remark this answers is not in the document]
Comment (marcokuchla) [reply kpis/src/main/java/objective/taskboard/followup/kpi/IssueKpi.java]: This avoids NPE when the properties configuration is wrong.

[the remark this answers is not in the document]
Comment (herbert-venancio) [reply kpis/src/main/java/objective/taskboard/followup/kpi/IssueKpi.java]: Can you "default" the properties configuration?
example: `ClusterAlgorithmProperties`

**Pull request merged**